Allow setting audodoc default options in conf

LewisHaley opened the pull request:
Subject: Set global default values for directives in config

### Feature or Bugfix
- Feature

### Purpose
- In RST files, you can set, for example: `:special-members: __init__, __iter__`. In config, you can set this flag, but only to include *all* special members, by including it in a list `autodoc_default_flags`. This PR make it so you can set the same values in global config as you can individually in RST files.

### Detail
- I wanted to write a test for this but I couldn't get the existing setup in `test_autodoc` to work for me, because `autodoc.directives.process_documenter_options` wants a `options` object which in the unit tests is mocked to a `Struct`, which doesn't work with `process_documenter_options`.
- I have managed to use this code to produce the docs for my own project, and it worked as expected using
```python
autodoc_default_values = {
    'members': None,
    'undoc-members': None,
    'special-members': None,
    'show-inheritance': None,
    'member-order': 'bysource',
    'exclude-members': '__dict__,__weakref__,__module__',
}
```
in conf.py

### Relates
- Issue: https://github.com/sphinx-doc/sphinx/issues/4075 (see final comment)
- Overlaps/replaces https://github.com/sphinx-doc/sphinx/pull/4076

LewisHaley committed:
test_autodoc: fix mutable function default argument in do_autodoc

Setting mutable types as default arguments is bad practice because the
value is only initialised once. This means that defaults arguments of
lists and dictionaries which are modified during code execution *stay*
modified between calls.

In this case, the `options` dictionary accumulated options as more and
more test cases were executed. Without this change, the tests added in
the next commit do not pass.

See: https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument

tk0miya commented:
As discussed in #4076, it would be better to enhance `autodoc_default_flags` to accept both list and dict:
```
autodoc_default_flags = [
    'members', 'undoc-members', ...
]

or

autodoc_default_flags = {
    'members': None,
    'undoc-members': None,
    ...
    'exclude-members': '__dict__,__weakref__,__module__',
}
```

Since 1.8, Sphinx provides `config-inited` event. I think it is good for conversion.
For example, HTML builder installs a converter for `html_js_files`:
https://github.com/sphinx-doc/sphinx/blob/8ddca8c62051d06654b22746a6f2d07a2e0ce750/sphinx/builders/html.py#L1605-L1620

tk0miya commented:
I just made #5223 for better testing (I'll merge it soon). I hope this helps this case.

LewisHaley commented:
I didn't like the idea of overloading `autodoc_default_flags` because (to me, at least) "flags" implies on/off only (which is how the option currently behaves - if "members" is present then "members = True", else "members = False"). I think the idea of giving arguments to "flags" is perhaps more confusing than just adding a new option. The old option could always be deprecated eventually.

But I'm happy to follow your suggestion of using the converter function with the original option name.
Thanks for the unittest - I'll try to follow it's example.

tk0miya commented:
OMG, I just made #5277 as an alternative of this PR. Okay, let's discuss what is the best way to configure the default settings. I'll reconsider this later (It's time to sleep in japan 😄 ).

@shimizukawa Please let me know your opinion.

LewisHaley committed:
autodoc: allow specifying values to global arguments

Previously, users could specify a *list* of flags in their config files.
The flags were directive names that would otherwise be present in the
.rst files. However, as a list, it was not possible to specify values
to those flags, which *is* possible in .rst files.

For example, in .rst you could say

  :special-members: __init__, __iter__

And this would cause autodoc to generate documents for these methods that
it would otherwise ignore.

This commit changes the config option to instead accept a dictionary.
This is a dictionary whose keys can contain the same flag-names as before,
but whose values can contain the arguments as seen in .rst files.

The old list is still supported, for backwards compatibility, but the data
is transformed into a dictionary when the user's config is loaded.

LewisHaley commented:
I've made the changes to use `config-inited` and the conversion hook.
I found a latent bug in the autodoc tests when I added some new tests, which I've fixed: https://github.com/sphinx-doc/sphinx/pull/5215/commits/50ad4f184d5341c8ed7588c6cfd5b39b7d8a26f7.

I've added a load of tests for this, partly because the different combinations of options can be quite surprising. I found testing the diff between option on/off was the best for showing what the *full* effect is of each of the options/values.

shimizukawa commented:
+1 for `autodoc_default_values`. `autodoc_default_options` is better to me because `:members:` is a option of `autodoc` directive.

tk0miya commented:
Okay, let's make it an alias of `autodoc_default_flags` on next step.

tk0miya commented on sphinx/ext/autodoc/__init__.py:
```diff
@@ -41,6 +42,8 @@
     from docutils.utils import Reporter  # NOQA
     from sphinx.application import Sphinx  # NOQA
+    from sphinx.config import Config  # NOQA
+    from sphinx.config import Config  # NOQA
```
It seems duplicated.

LewisHaley replied:
fixed

tk0miya commented on tests/test_autodoc.py:
```diff
@@ -1411,22 +1417,248 @@ def test_partialmethod(app):
+
+    # with :special-members:
+    # Note that :members: must be *on* for :special-members: to work.
+    # TODO Is this a bug?
```
AFAIK, it's well-known behavior.

LewisHaley replied:
removed TODO comment

tk0miya commented on tests/test_autodoc.py:
```diff
@@ -1411,22 +1417,248 @@ def test_partialmethod(app):
+
+
+@pytest.mark.sphinx('html', testroot='ext-autodoc')
+@pytest.mark.parametrize('target, options, expected_diff', [
```
I feel this parameterized test is hard to understand...
Could you change this to simple testcases?

In addition, is doing diff important for this case? We've already done testing for each options. So I think the main purpose of this case is `autodoc_default_flags` modifies the options of each autodoc results.

LewisHaley replied:
Done. I found the diffs useful for myself as I wasn't completely familiar with what each option did but, as you say, there are tests for this already. I've left the tests quite verbose but they don't use `parametrize` or look at the diffs.

tk0miya reviewed (changes requested):
LGTM with nits

LewisHaley committed:
autodoc: add 'exclude-members' to user global options

As the previous commit explains, it is now possible to specify arguments
to the global options in config files. This means that we can now include
the `exclude-members` option in this global configuration.

Previously, there was no point including this option because it makes no
sense without arguments. Including this option means users have the
flexibility of explicitly including which special methods they want using
(e.g.):

  :special-members: __init__, __iter__

or explicitly excluding which special-members (or other members) they want
using (e.g.):

  :exclude-members: __weakref__, __hash__

LewisHaley commented:
Re. future work about making an alias for the option so its name makes more sense, could I suggest as well that the accepted types for the dict is more tightly defined. For example, `:undoc-members:` would make more sense as only allowing True/False, while `:exclude-members:` only make sense as a list of strings (if the list is empty, nothing is excluded).

`:members:`, `:special-members:` are more difficult because they need a "say yes to everything" as well as a limiting list. You could use `{'members': ["*"]}` (meaning glob) to include everything? This way, the values in the dictionary are *always* a list of strings.

Just some thoughts I've had. :)

tk0miya commented:
Just merged!
Thank you for contribution!
